### ISSUE TITLE
Parallelism issues

### ISSUE THREAD

**Issue by essen** [emqx/quic#229]
Hello, I wanted to ask if you have done some (recent) testing with many connections in parallel. In Cowboy if I run the tests sequentially they run mostly fine, but if I run them in parallel I always get a few that seem to get stuck. I am not yet sure where exactly, but it happens soon after the connection has been opened and both client and server have started their unidi streams (3 each). So it likely happens either when the client starts the first bidi stream, sends data on it, or waits for the response; or when the server receives this bidi stream, the data and/or sends the response.

Another issue I had that I've seen resolved is msquic refusing connections when overloaded; here increasing the number of acceptors and having the client retry seems to help.

**Comment by qzhuyan** [comment]
> Another issue I had that I've seen resolved is msquic refusing connections when overloaded; here increasing the number of acceptors and having the client retry seems to help.

Msquic refuse conn when running low mem. You could tune it look for `retry_memory_limit`

for perf test, No, I haven't done it for version 0.300.0. we did mostly for 0.100.x.
I plan to do it soon but am busy with other urgent things. 

> if I run the tests sequentially they run mostly fine, but if I run them in parallel I always get a few that seem to get stuck.

Do you have process per stream or one process handles both stream and connecion?

Have you checked any messages unconsumed in the message queue in the form of `{quic, _, _, _}`

would be helpful if you could give me some sample code. 

last kill, if you could compile quicer with envvar `QUIC_LOGGING_TYPE=stdout` then run your test. lots of debug will be printed and you could share the debug printing here so I could take a look into what happened.  (would be helpful if I know which stream/connecion is blocked, src port will also help).

**Comment by essen** [comment]
The failures are unreliable and there's 80 tests running in parallel so logging is a bit difficult. That's why I'm fishing for information. I'll keep you informed when I make progress and figure out where it's coming from exactly. I got a few ideas for that but week-end comes first. Cheers.

**Comment by qzhuyan** [comment]
Dont worry about the huge logs, I have tech to handle big ones.

**Comment by essen** [comment]
The problem came from me calling in this order accept, handshake and controlling_process, when I should have done accept, controlling_process and then in that other process, handshake. So messages were received by the wrong process. It's a mistake that built on early misunderstanding I had with the API.

I still have one issue to investigate but will reopen if it turns out a QUIC related problem.

Cheers!

**Comment by essen** [comment]
For the record (and future reference) the other issue I am running into is not a bug, it's that QUIC streams may fail to (re)send all the data after the stream is reset. So it makes it difficult to send data then reset a stream. There is currently a draft RFC to address this issue at https://www.ietf.org/id/draft-ietf-quic-reliable-stream-reset-03.html

**Comment by qzhuyan** [comment]
@essen reset stream is like abort the sending.

If you want to ensure data get delivered then shutdown the stream,

I think you could shutdown stream gracefully OR send the last piece of data with FIN flag.

**Comment by essen** [comment]
Yes I shut down the stream, the problem happens when you shut down the sending part of a stream. It's a protocol issue, not much I can do except NOT shut it down. In this scenario it's not a big problem anyway as the shut down error code is descriptive enough as to what happened.

**Comment by qzhuyan** [comment]
may relates to #243
